thorlcr/master/thgraphmanager.cpp

```diff
@@ -1463,63 +1464,156 @@ void thorMain(ILogMsgHandler *logHandler, const char *wuid, const char *graphNam
                         else
                         {
-                            JobNameScope activeJobName(currentWuid.str());
-                            saveWuidToFile(currentWuid);
-                            PROGLOG("Executing: wuid=%s, graph=%s", currentWuid.str(), currentGraphName.str());
-
+                            SessionId agentSessionID = workunit->getAgentSession();
+                            if (agentSessionID <= 0)
                             {
-                                Owned<IWorkUnit> wu = &workunit->lock();
-                                publishPodNames(wu, currentGraphName, nullptr);
+                                WARNLOG("Discarding job with invalid sessionID: wuid=%s, graph=%s (sessionID=%" I64F "d)", currentWuid.str(), currentGraphName.str(), agentSessionID);
+                                currentWuid.clear();
                             }
-                            SocketEndpoint dummyAgentEp;
-                            jobManager->execute(workunit, currentWuid, currentGraphName, dummyAgentEp);
-
-                            Owned<IWorkUnit> w = &workunit->lock();
-                            if (!multiJobLinger && lingerPeriod)
-                                w->setDebugValue(instance, "1", true);
-
-                            if (jobManager->queryExitException())
+                            else if (querySessionManager().sessionStopped(agentSessionID, 0))
                             {
-                                // NB: exitException has already been relayed.
-                                jobManager->clearExitException();
+                                WARNLOG("Discarding agentless job: wuid=%s, graph=%s", currentWuid.str(), currentGraphName.str());
+                                currentWuid.clear();
                             }
                             else
                             {
-                                switch (w->getState())
+                                SCMStringBuffer jobVersion;
+                                bool mismatch = false;
+                                if (imagePattern) // null if failed to parse (see above), will continue without looking at versioning
                                 {
-                                    case WUStateRunning:
-                                        w->setState(WUStateWait);
-                                        break;
-                                    case WUStateAborting:
-                                    case WUStateAborted:
-                                    case WUStateFailed:
-                                        break;
-                                    default:
-                                        w->setState(WUStateFailed);
-                                        break;
+                                    workunit->getDebugValue("platformVersion", jobVersion);
+                                    if (0 == runtimePlatformVersion.length()) // 1st job
```

Comment:
Is it guaranteed that the job that started a thor option will be run on that thor?
Better would be if there was another environment variable (or other mechanism) to get at the current version - and not rely on the first job happening to match.

Reply:
Having an image: xxxx environment variable might allow it to fall out.